BBQ OSR should count try exceptions only in IPInt OSR

danlliu opened the pull request:
#### 4d8c1e151637febc0fa05fcbb3c709bae3c27c3a
<pre>
BBQ OSR should count try exceptions only in IPInt OSR
<a href="https://bugs.webkit.org/show_bug.cgi?id=286960">https://bugs.webkit.org/show_bug.cgi?id=286960</a>
<a href="https://rdar.apple.com/144077240">rdar://144077240</a>

Reviewed by Yijia Huang.

Previously, we fixed the issue of an incorrect stackmap exception slot count by
updating the definition of implicitSlots. However, this leads to issues later on
in BBQ, since this calculation leads to incorrect offsets and indexing out of bounds.
This patch fixes that issue by only counting the exception slot when calculating
the IPInt-&gt;BBQ stackmap, and otherwise using the original counting for implicit slots.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::ControlData::implicitSlots const):
(JSC::Wasm::BBQJITImpl::BBQJIT::makeStackMap):

Canonical link: <a href="https://commits.webkit.org/289960@main">https://commits.webkit.org/289960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e62e7c9a5d3b6c12be48598f3aca2790fb76c01

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93299 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39096 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16044 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68139 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25862 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91344 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6308 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79905 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48508 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6080 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34316 "Found 4 new test failures: accessibility/nested-custom-element-accname.html editing/undo/redo-reapply-edit-command-crash.html http/tests/security/file-system-access-via-dataTransfer.html imported/w3c/web-platform-tests/xhr/request-content-length.any.worker.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38204 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81141 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76462 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35200 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95142 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87118 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15517 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11372 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77007 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15773 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76256 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20650 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19070 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8554 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13845 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15533 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20840 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109611 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15274 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26359 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18723 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17056 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me, is that possible to add a test?

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3301,6 +3299,12 @@ B3::ValueRep BBQJIT::toB3Rep(Location location)
 StackMap BBQJIT::makeStackMap(const ControlData& data, Stack& enclosingStack)
 {
     unsigned numElements = m_locals.size() + data.enclosedHeight() + data.argumentLocations().size();
+    if (Options::useWasmIPInt()) {
```
I seem to recall that the LLInt->BBQ stackmap is re-used for BBQ->OMG, in which case this would be wrong. But then the code would already be very wrong, so I must be misremembering. Can you double check?

danlliu replied:
OMG uses a different stackmap completely, in WasmIRGeneratorHelpers.h; so I don't believe this should affect BBQ->OMG.

danlliu committed:
BBQ OSR should count try exceptions only in IPInt OSR
https://bugs.webkit.org/show_bug.cgi?id=286960
rdar://144077240

Reviewed by Yijia Huang.

Previously, we fixed the issue of an incorrect stackmap exception slot count by
updating the definition of implicitSlots. However, this leads to issues later on
in BBQ, since this calculation leads to incorrect offsets and indexing out of bounds.
This patch fixes that issue by only counting the exception slot when calculating
the IPInt->BBQ stackmap, and otherwise using the original counting for implicit slots.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::ControlData::implicitSlots const):
(JSC::Wasm::BBQJITImpl::BBQJIT::makeStackMap):

Canonical link: https://commits.webkit.org/289960@main
